### PR TITLE
feat(setup): remove zx & use execa

### LINKED ISSUE #340: 🛠 Tooling: Consider execa vs zx
### Bug Report Checklist

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have pulled the latest `main` branch of the repository.
- [X] I have [searched for related issues](https://github.com/JoshuaKGoldberg/template-typescript-node-package/issues?q=is%3Aissue) and found none that matched my issue.

### Overview

Forking from @navin-moorthy's https://github.com/JoshuaKGoldberg/template-typescript-node-package/issues/334#issuecomment-1490392645: 

> What's your thought on using `execa` [release v7.1.0](https://github.com/sindresorhus/execa/releases/tag/v7.1.0) with script interface ?
> 
> [`execa` vs `zx` comparsion](https://github.com/sindresorhus/execa/blob/main/docs/scripts.md?rgh-link-date=2023-03-30T14%3A19%3A35Z).

### Additional Info

I'm not tied to either tool. This repo already uses zx right now and it works great. So I'd need to see some strong reason to use execa instead to switch. Is there a good motivation,  @navin-moorthy? 

_(personal familiarity notwithstanding 😄)_

### PR DESCRIPTION
**Pull request by navin-moorthy**
## PR Checklist

- [x] Addresses an existing open issue: fixes #340
- [x] That issue was marked as [`status: accepting prs`](https://github.com/JoshuaKGoldberg/template-typescript-node-package/issues?q=is%3Aopen+is%3Aissue+label%3A%22status%3A+accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/JoshuaKGoldberg/template-typescript-node-package/blob/main/.github/CONTRIBUTING.md) were taken

## Overview

- Maintain the clack prompt till the end with intro, spinner & outro
- Add confirm clack promt for `skip-api`
- Improve some node commands with well tested packages which are well tested
- Added try catch block to every command to catch the errors & the message in granular
- Reorganize catching errors without having to use `caughtError`
- Added `{cause: error}` to get the complete stack trace in the final catch block
- Added restoring the half completed changed on error
- Remove dependencies only on successfull setup run
- Update test file with `execa`

### CODE REVIEW

**Comment by navin-moorthy** [comment]
Images of how the setup would look like now,

<details><summary>
Check the preview of how this command would run in terminal,

```sh
pnpm run setup --repository "new-repository-test" --title "New Title Test" --owner "NewOwnerTest" --description "New Description Test" --skip-api
```
</summary>

image
</details> 

<details><summary>
Check the preview of how this command would run in terminal `without any errors`,

```sh
pnpm run setup
```
</summary>

image

</details> 


<details><summary>
Check the preview of how this command would run in terminal `with errors`,

```sh
pnpm run setup
```
</summary>

image

</details>

File: script/setup.js [hunk 1]
```diff
@@ -197,191 +234,296 @@ try {
...
+			"./CHANGELOG.md",
+			prettier.format(`# Changelog`, { parser: "markdown" })
+		);
+	} catch (error) {
+		s.stop(chalk.red`❌ Error clearing CHANGELOG.md.`);
 
-	console.log();
-	console.log(chalk.gray`Generating all-contributors table in README.md...`);
+		throw new Error("Could not empty 'CHANGELOG.md'. ", { cause: error });
+	}
```
Comment (JoshuaKGoldberg) [inline script/setup.js:301]: [Refactor] There are a lot of similar chunks of code that roughly:

1. `s.start(...)`
2. `try`/`catch`
3. `s.stop(...)`

I'd request that a general helper function be made for them (in #351) to help standardize. It's a good change!

File: script/setup.js [hunk 2]
```diff
@@ -197,191 +234,296 @@ try {
...
-		await $`pnpm remove @clack/prompts all-contributors-cli chalk octokit npm-user replace-in-file title-case -D`;
+		await $`pnpm remove @clack/prompts all-contributors-cli chalk octokit npm-user replace-in-file titleize read-pkg-up load-json-file git-remote-origin-url git-url-parse execa -D`;
 	} catch (error) {
-		console.log("Error uninstalling packages:", error);
-		caughtError = error;
+		s.stop(
+			chalk.red`❌ Error removing devDependency packages only used for setup`
+		);
+
+		throw new Error("Could not uninstalling packages. ", {
```
Comment (JoshuaKGoldberg) [inline script/setup.js:492]: [Typo]

```suggestion
		throw new Error("Could not uninstall packages. ", {
```

Aside: love the `cause: error`. Great new feature!

File: script/setup.js [hunk 3]
```diff
@@ -197,191 +234,296 @@ try {
...
+	console.log(chalk.gray`git commit -m "chore: hydrated repo"`);
+	console.log(chalk.gray`git push`);
+	console.log();
 	console.log(chalk.greenBright`See ya! 👋`);
-}
+	console.log();
+
+	exitCode = 0;
+} catch (error) {
+	s.start(`Restoring the changes done... `);
```
Comment (JoshuaKGoldberg) [inline script/setup.js:512]: This feature of using `git` to restore changes is a great idea that should also be a separate PR. I'd want it to prompt the user to ask if they want to do that. Sometimes it mostly works & that alone is enough to setup (or debug why it failed).

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
🙌 this is a lot of great stuff, thanks so much for sending it in @navin-moorthy!

Because this is a lot of great stuff 😄 requesting it be split up into more granular PRs. It's hard to review lots of changes together - and slower to get each of them merged in. I can review more closely once they're split up. Also I filed #350 on myself to document this in the contributing guidelines.

**Comment by navin-moorthy** [comment]
Sure, I also felt the same way after finishing the PR. I will split the PR 👍.

File: script/setup.js (hunk 3 above)
Comment (navin-moorthy) [reply script/setup.js:512]: Right, we can move this to separate PR & ask this the prompt.

File: script/setup.js [hunk 4]
```diff
@@ -1,27 +1,37 @@
...
 import npmUser from "npm-user";
 import { Octokit } from "octokit";
 import prettier from "prettier";
+import { readPackageUp } from "read-pkg-up";
 import replace from "replace-in-file";
-import { titleCase } from "title-case";
+import titleize from "titleize";
 
-let caughtError;
+const s = spinner();
```
[the remark this answers is not in the document]
Comment (navin-moorthy) [reply script/setup.js:26]: Yes, I will move the Spinner logic to a separate PR.

File: script/setup-test-e2e.js [hunk 5]
```diff
@@ -25,10 +22,11 @@ for (const search of [
 	`/JoshuaKGoldberg/`,
 	"template-typescript-node-package",
 ]) {
-	const grepResult =
-		await $`grep --exclude script/setup.js --exclude script/setup-test-e2e.js --exclude-dir node_modules -i ${search} *.* **/*.*`;
+	const { stdout } = await $({
+		shell: true,
```
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply script/setup-test-e2e.js:28]: Hmm yeah when I try it locally, as-is or with `"` quotes around the glob patterns, it runs forever. Fun! I guess this is fine, but I'll add an issue about looking into getting this to work without the `shell: true`. Thanks for looking into it!

**Review by JoshuaKGoldberg — APPROVED** [review]
Super, thanks @navin-moorthy!

File: script/setup-test-e2e.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (navin-moorthy) [reply script/setup-test-e2e.js:28]: Sure 👍. I have merged this branch with main now.

File: script/setup-test-e2e.js [hunk 6]
```diff
@@ -25,10 +24,11 @@ for (const search of [
...
```
Comment (JoshuaKGoldberg) [inline script/setup-test-e2e.js:28]:
```suggestion
		// Todo: it'd be nice to not have to use this... but we haven't figured out how yet.
		// https://github.com/JoshuaKGoldberg/template-typescript-node-package/issues/357
		shell: true,
```

File: script/setup-test-e2e.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (navin-moorthy) [reply script/setup-test-e2e.js]: Added issue with `execa` for resolutions - https://github.com/sindresorhus/execa/issues/555

**Comment by navin-moorthy** [comment]
Seems, [Knip v2.4.0](https://github.com/webpro/knip/releases/tag/2.4.0) introduced some bugs causing the lint:knip to fail with below Unlisted dependencies

image

File: knip.jsonc [hunk 7]
```diff
@@ -1,6 +1,7 @@
 {
 	"$schema": "https://unpkg.com/knip@next/schema.json",
 	"entry": ["src/index.ts!", "script/setup*.js"],
+	"ignoreBinaries": ["cat", "gh"],
```
Comment (JoshuaKGoldberg) [inline knip.jsonc:4]: fyi @navin-moorthy - this fixes the `lint:knip` CI failure. https://github.com/webpro/knip/issues/116 & https://github.com/webpro/knip/issues/117 have a bit more context.

**Comment by JoshuaKGoldberg** [comment]
> Seems, [Knip v2.4.0](https://github.com/webpro/knip/releases/tag/2.4.0) introduced some bugs causing the lint:knip to fail with below Unlisted dependencies
> 
> image
> 
> References: [webpro/knip#113](https://github.com/webpro/knip/pull/113)

That awkward moment when I'm the one who sent that PR 😂

**Comment by navin-moorthy** [comment]
Ha ha ha, yes seems some edge cases needs to be handled.
